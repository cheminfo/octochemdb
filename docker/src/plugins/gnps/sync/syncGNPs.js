import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import debugLibrary from '../../../utils/Debug.js';
import createIndexes from '../../../utils/createIndexes.js';
import { shouldUpdate } from '../../../utils/shouldUpdate.js';

import { parseGNPs } from './utils/parseGNPs.js';

const debug = debugLibrary('syncGNPs');
/**
 * @description Synchronize GNPS collection with the GNPS database
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns gnps collections
 */
export async function sync(connection) {
  try {
    let options = {
      collectionSource: process.env.GNPS_SOURCE,
      destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/gnps/full`,
      collectionName: 'gnps',
      filenameNew: 'gnps_full',
      extensionNew: 'json',
    };
    // Get lastFile (path), sources, progress, logs,lastDocumentImported and collection gnps
    let sources;
    let lastFile;
    if (process.env.NODE_ENV === 'test') {
      lastFile = `${process.env.GNPS_SOURCE_TEST}`;
      sources = [process.env.GNPS_SOURCE_TEST];
    } else {
      lastFile = await getLastFileSync(options);
      sources = [lastFile.replace(`${process.env.ORIGINAL_DATA_PATH}`, '')];
    }

    const progress = await connection.getProgress(options.collectionName);

    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );
    let isTimeToUpdate = await shouldUpdate(
      progress,
      sources,
      lastDocumentImported,
      process.env.GNPS_UPDATE_INTERVAL,
      connection,
    );
    // define counters
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    if (isTimeToUpdate) {
      const collection = await connection.getCollection(options.collectionName);

      const logs = await connection.getImportationLog({
        collectionName: options.collectionName,
        sources,
        startSequenceID: progress.seq,
      });
      // create temporary collection
      const temporaryCollection = await connection.getCollection(
        `${options.collectionName}_tmp`,
      );
      debug.info(`Start importing GNPs`);
      // set progress to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // parse GNPs
      for await (const entry of parseGNPs(lastFile, connection)) {
        counter++;
        // if test mode is enabled, stop after 20 entries
        if (process.env.NODE_ENV === 'test' && counter > 20) break;

        if (Date.now() - start > Number(process.env.DEBUG_THROTTLING)) {
          debug.trace(
            `Processing: counter: ${counter} - imported: ${imported}`,
          );
          start = Date.now();
        }
        // insert entry in temporary collection
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        imported++;
      }
      // rename the temporary collection to the final collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });
      // update the logs and progress
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // create indexes on the collection
      await createIndexes(collection, [
        { 'data.ocl.idCode': 1 },
        { 'data.ocl.noStereoTautomerID': 1 },
        { 'data.spectrum.msLevel': 1 },
        { 'data.spectrum.ionSource': 1 },
        { 'data.spectrum.precursorMz': 1 },
        { 'data.spectrum.adduct': 1 },
        { 'data.spectrum.ionMode': 1 },
        { 'data.spectrum.data.x': 1 },
        { 'data.spectrum.data.y': 1 },
        { 'data.spectrum.numberOfPeaks': 1 },
        { 'data.em': 1 },
        { 'data.mf': 1 },
        { _seq: 1 },
      ]);
      debug.trace(`${imported} compounds processed`);
      debug.info(`GNPs imported`);
    } else {
      debug.info(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      await debug.fatal(e.message, {
        collection: 'gnps',
        connection,
        stack: e.stack,
      });
    }
  }
}
