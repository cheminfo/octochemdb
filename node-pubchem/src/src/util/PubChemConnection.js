'use strict';

const delay = require('delay');
const { MongoClient } = require('mongodb');
const debug = require('debug')('PubChemConnection');

function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function getProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.updateOne({ _id: progress._id }, { $set: progress });
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    debug('Starting new database construction.');
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  } else {
    if (progress.state === 'update') {
      debug('First importation has been completed. Should only update.');
    } else {
      debug(`Continuing first importation from ${progress.seq}.`);
    }
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};

module.exports = PubChemConnection;
