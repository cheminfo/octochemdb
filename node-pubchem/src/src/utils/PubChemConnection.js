import Debug from 'debug';
import delay from 'delay';
import { MongoClient } from 'mongodb';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      console.log(e);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
