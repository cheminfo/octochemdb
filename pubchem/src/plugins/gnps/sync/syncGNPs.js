import md5 from 'md5';

import getLastDocumentImported from '../../../sync/http/utils/getLastDocumentImported.js';
import getLastFileSync from '../../../sync/http/utils/getLastFileSync.js';
import Debug from '../../../utils/Debug.js';

import { parseGNPs } from './utils/parseGNPs.js';

const debug = Debug('syncGNPs');
/**
 * @description Synchronize GNPS collection with the GNPS database
 * @param {*} connection MongoDB connection
 * @returns {Promise} returns gnps collections
 */
export async function sync(connection) {
  try {
    let options = {
      collectionSource: process.env.GNPS_SOURCE,
      destinationLocal: `${process.env.ORIGINAL_DATA_PATH}/gnps/full`,
      collectionName: 'gnps',
      filenameNew: 'gnps_full',
      extensionNew: 'json',
    };
    // Get lastFile (path), sources, progress, logs,lastDocumentImported and collection gnps
    const lastFile = await getLastFileSync(options);
    const sources = [lastFile.replace(process.env.ORIGINAL_DATA_PATH, '')];

    const progress = await connection.getProgress(options.collectionName);
    const collection = await connection.getCollection(options.collectionName);

    const logs = await connection.getImportationLog({
      collectionName: options.collectionName,
      sources,
      startSequenceID: progress.seq,
    });

    const lastDocumentImported = await getLastDocumentImported(
      connection,
      progress,
      options.collectionName,
    );
    // define counters
    let counter = 0;
    let imported = 0;
    let start = Date.now();
    if (
      lastDocumentImported === null ||
      md5(JSON.stringify(sources)) !== progress.sources ||
      progress.state !== 'updated'
    ) {
      // create temporary collection
      const temporaryCollection = await connection.getCollection(
        `${options.collectionName}`,
      );
      debug(`Start parsing: ${lastFile}`);
      // set progress to updating
      progress.state = 'updating';
      await connection.setProgress(progress);
      // parse GNPs
      for await (const entry of parseGNPs(lastFile, connection)) {
        counter++;
        // if test mode is enabled, stop after 20 entries
        if (process.env.TEST === 'true' && counter > 20) break;

        if (
          Date.now() - start >
          Number(process.env.DEBUG_THROTTLING || 10000)
        ) {
          debug(`Processing: counter: ${counter} - imported: ${imported}`);
          start = Date.now();
        }
        // insert entry in temporary collection
        entry._seq = ++progress.seq;
        await temporaryCollection.updateOne(
          { _id: entry._id },
          { $set: entry },
          { upsert: true },
        );
        imported++;
      }
      // rename the temporary collection to the final collection
      await temporaryCollection.rename(options.collectionName, {
        dropTarget: true,
      });
      // update the logs and progress
      logs.dateEnd = Date.now();
      logs.endSequenceID = progress.seq;
      logs.status = 'updated';
      await connection.updateImportationLog(logs);
      progress.sources = md5(JSON.stringify(sources));
      progress.dateEnd = Date.now();
      progress.state = 'updated';
      await connection.setProgress(progress);
      // create indexes on the collection
      await collection.createIndex({ _id: 1 });
      await collection.createIndex({ 'data.ocl.id': 1 });
      await collection.createIndex({ 'data.ocl.noStereoID': 1 });
      await collection.createIndex({ 'spectralData.msLevel': 1 });
      await collection.createIndex({ 'spectralData.ionSource': 1 });
      await collection.createIndex({ 'spectralData.precursorMz': 1 });
      await collection.createIndex({ 'spectralData.adduct': 1 });
      await collection.createIndex({ 'spectralData.ionMode': 1 });
      await collection.createIndex({ 'spectralData.spectrum': 1 });

      debug(`${imported} compounds processed`);
    } else {
      debug(`file already processed`);
    }
  } catch (e) {
    if (connection) {
      debug(e, { collection: 'gnps', connection });
    }
  }
}
