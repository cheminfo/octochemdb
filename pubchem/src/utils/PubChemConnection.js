import delay from 'delay';
import { MongoClient } from 'mongodb';

import Debug from './Debug.js';

const debug = Debug('PubChemConnection');

export function PubChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

PubChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

PubChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

PubChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

PubChemConnection.prototype.getLogs = async function getCollection(
  collectionName,
) {
  const logsCollection = await this.getLogsCollection();
  const _id = `${collectionName}`;
  let logs = await logsCollection.find({ _id }).next();
  if (logs === null) {
    let sources = [];
    logs = {
      _id,
      sources,
      sourcesHash: md5(JSON.stringify(sources)),
      dataStart: 0,
      dataEnd: 0,
      startSequenceID: 123,
      endSeqneceID: 123,
      status: 'updating',
    };
    await logsCollection.insertOne(logs);
  }
  return logs;
};
PubChemConnection.prototype.setLogs = async function setLogs(logs) {
  const collection = await this.getLogsCollection();
  await collection.replaceOne({ _id: logs._id }, logs);
};
PubChemConnection.prototype.getLogsCollection = async function getCollection() {
  return (await this.getDatabase()).collection('logs');
};

PubChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

PubChemConnection.prototype.setProgress = async function setProgress(progress) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

PubChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'import',
      seq: 0,
      date: new Date(),
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

PubChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

PubChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
