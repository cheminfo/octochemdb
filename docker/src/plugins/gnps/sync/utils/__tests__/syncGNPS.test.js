import { PubChemConnection } from '../../../../../utils/PubChemConnection.js';
import { sync } from '../../syncGNPs';
jest.setTimeout(300000);
test('syncGNPs', async () => {
  const connection = new PubChemConnection();
  await sync(connection);
  const collection = await connection.getCollection('gnps');
  const collectionEntry = await collection
    .find({ _id: 'CCMSLIB00000001547' })
    .limit(1);
  const result = await collectionEntry.next();
  expect(result).toMatchSnapshot();
  const emptySmilesEntry = await collection
    .find({ _id: 'CCMSLIB00000001548' })
    .limit(1);
  const emptySmiles = await emptySmilesEntry.next();
  // expect(bronzeSpectrum) to be null
  expect(emptySmiles).toBeNull();

  if (connection) {
    await connection.close();
  }
});
