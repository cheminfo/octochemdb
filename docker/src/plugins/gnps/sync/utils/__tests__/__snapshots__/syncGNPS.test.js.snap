// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`syncGNPs 1`] = `
{
  "_id": "CCMSLIB00000001547",
  "data": {
    "ocl": {
      "coordinates": "!BdQXYo\`{VLRdQyf[ASjdY_ae^}cYf_mFiBXeayafAmgY?XNGiHV[ihV{f}dei[c[vBS[azVY^rWGfD\\\\EfzZD^ykxYjRxImfGiyazIm\`Efb[E^Yc[qJTzVSaYnjUG~YkYqJXzVS\`DY?fZVL]YqedyAF_yfecEidZeNedyAF\\\\dN?fZVSjeN?f[~YjeN",
      "idCode": "gfOEPX@@DE@qTzJmPy}~@ZKARx|aXbTjL^rJIQQSSPyQJISJIQPpqSPsPiIIZIYQJZXqIVgQFbuSULuUMUUMSMUMULtuUUM@A@A@zlt@",
      "noStereoTautomerID": "gfOEPX@@DE@qTzJmPy}~@ZKARx|aTfTjL^rJIQQSSPyQJISJIQPpqSQRpiIIZIYQJZXqIVgQFcUSUUUUUUUUUUUUUUUUUUU@A@@AjV@X\`XpXXXxXTXtXLXlX\\\\X|XBXbXRXrXJXjXZXzXFXfXVXvXNXnX^X~XAXaXQXqXIXiXYXEXeXUXuXMXmX]X}XCXcXSXsXkX[XWXoX_XhxQA\`Aidq|L_CCjfOacxX",
    },
    "spectrum": {
      "adduct": "M+H",
      "data": {
        "x": [
          361.84436,
          375.751953,
          427.667358,
          443.266113,
          446.267273,
          447.747498,
          455.25,
          456.107544,
          469.872314,
          471.062195,
          476.975159,
          478.891113,
          479.975952,
          487.210388,
          491.191956,
          495.653992,
          503.028198,
          504.344543,
          510.176514,
          515.922852,
          520.973389,
          523.168945,
          530.991211,
          538.003174,
          539.217773,
          540.672852,
          554.117432,
          556.030396,
          557.288818,
          564.123047,
          564.94873,
          566.439941,
          571.33374,
          575.219238,
          577.102905,
          582.110107,
          585.237061,
          598.172485,
          599.352783,
          600.382812,
          602.267578,
          622.208984,
          623.023193,
          641.235107,
          659.420898,
          668.33252,
          680.219727,
          681.987793,
          691.648682,
          693.22583,
          696.332397,
          711.744873,
          714.071777,
          715.578979,
          724.081909,
          735.806396,
          769.275635,
          787.432617,
          828.517456,
          830.409302,
          831.306763,
          838.518677,
          839.455811,
          847.433594,
          848.125854,
          851.380859,
          852.370605,
          853.270508,
          865.5979,
          866.295654,
          868.372192,
          877.137451,
          883.442261,
          888.167114,
          889.282959,
          893.467896,
          899.010132,
          909.424438,
          910.515381,
          915.217773,
          919.39624,
          922.205688,
          925.063721,
          931.132812,
          932.351929,
          933.524048,
          935.493286,
          936.552002,
          937.588623,
          938.471069,
          939.61792,
          946.257812,
          949.370239,
          951.551758,
          953.396606,
          954.491577,
          963.686768,
          964.524658,
          965.192139,
          982.221924,
        ],
        "y": [
          0.031779292447709225,
          0.03321417406670565,
          0.08912354500780226,
          0.05811074356167179,
          0.030701496227059227,
          0.03831408603796092,
          0.09278857536006112,
          0.13785327583342816,
          0.11457340066943703,
          0.04117861725546517,
          0.0371865856226693,
          0.03648026285670935,
          0.04347155023459072,
          0.043013748441838896,
          0.07203576609206001,
          0.042128228974144664,
          0.040081200957982954,
          0.04763885455375443,
          0.050869627205460136,
          0.10302371544086966,
          0.03774510380982652,
          0.07707550982769648,
          0.04165342311480492,
          0.28900242373349133,
          0.3561645627404277,
          0.057390032739368206,
          0.09970269043573575,
          0.2804637662961088,
          0.06928503132018266,
          0.033970201027307224,
          0.045296217379987264,
          0.04651789416407355,
          0.08018725401328672,
          0.055102331780731255,
          0.03865155135947512,
          0.13351462284326307,
          0.048100580361872695,
          0.034119313611232105,
          1,
          0.1494618212924922,
          0.035504491035586894,
          0.031113517840535863,
          0.08363384751014685,
          0.08616222141125905,
          0.05489566697143186,
          0.08593593652512743,
          0.05804141929019794,
          0.031975493216031436,
          0.0381636654489139,
          0.04387703182245662,
          0.1585446088606883,
          0.04273906736618781,
          0.06603725460188903,
          0.0713739155002531,
          0.11577153336132465,
          0.0680450424643863,
          0.0549349071250963,
          0.04316809304625237,
          0.04202882058486141,
          0.13417908944531426,
          0.08540488644553532,
          0.03421741399539321,
          0.09418421682539309,
          0.25697330230745186,
          0.05184539902658259,
          0.32199162091918754,
          0.3621630742305987,
          0.057834754480898545,
          0.058463904944651764,
          0.14520426461990024,
          0.10205448364535796,
          0.039411502335443147,
          0.06440748021969254,
          0.05367660619758987,
          0.03243198700366111,
          0.03074596840121226,
          0.03745211066246536,
          0.31933113850073835,
          0.057251384196420514,
          0.037219285750723,
          0.1122647716288457,
          0.049524997939891936,
          0.056760882275614995,
          0.08074577220044393,
          0.17874805597738722,
          0.03296434508837537,
          0.04433614162033059,
          0.13489456824712925,
          0.08842768628281948,
          0.04627591321647616,
          0.10109440788570129,
          0.05046806963296068,
          0.11172979753388715,
          0.03923361363883101,
          0.7132303410100154,
          0.1621102308236639,
          0.34214536384124483,
          0.4161601416831148,
          0.1627223772208292,
          0.03550841505095334,
        ],
      },
      "instrument": "qTof",
      "ionMode": "Positive",
      "ionSource": "LC-ESI",
      "libraryQualityLevel": "Gold",
      "msLevel": 2,
      "numberOfPeaks": 100,
      "precursorMz": 981.54,
    },
  },
}
`;
